crates/wasmprinter/src/operator.rs

```diff
@@ -1286,3 +1341,450 @@ impl<'a> VisitOperator<'a> for PrintOperator<'_, '_, '_, '_> {
+            _ => None,
+        }
+    }
+
+    fn func_type_arity(&self, func_idx: u32) -> Option<(u32, u32)> {
+        match self.state.core.func_to_type.get(func_idx as usize) {
+            Some(Some(type_idx)) => self.sub_type_arity(*type_idx),
+            _ => None,
+        }
+    }
+
+    fn sub_type_arity(&self, type_idx: u32) -> Option<(u32, u32)> {
+        match self.state.core.types.get(type_idx as usize) {
+            Some(Some(SubType {
+                composite_type: CompositeType { inner: x, .. },
+                ..
+            })) => match x {
+                CompositeInnerType::Func(f) => {
+                    Some((f.params().len() as u32, f.results().len() as u32))
+                }
+                CompositeInnerType::Struct(s) => {
+                    Some((s.fields.len() as u32, s.fields.len() as u32))
+                }
+                _ => None,
+            },
+            _ => None,
+        }
+    }
+
+    fn control_stack_height(&self) -> u32 {
+        self.control.len() as u32
+    }
+
+    fn label_block(&self, depth: u32) -> Option<(BlockType, FrameKind)> {
+        let cur_depth = self.printer.nesting - self.operator_state.nesting_start;
+        if self.control.len() != cur_depth as usize + 1 {
+            return None;
+        }
+        match (self.control.len() - 1).checked_sub(depth as usize) {
+            Some(i) => Some((self.control[i].ty, self.control[i].kind)),
+            None => None,
+        }
+    }
+}
+
+// The define_visit_folded macro defines a visitor for each operator that will push the corresponding
+// folded instruction (or block entry) on the stack. The folded instructions will be printed to the
+// "real" printer in `finalize`.
+macro_rules! define_visit_folded {
+    ($(@$proposal:ident $op:ident $({ $($arg:ident: $argty:ty),* })? => $visit:ident ($($ann:tt)*) )*) => ($(
+        fn $visit(&mut self $( , $($arg: $argty),* )?) -> Self::Output {
+	    let arities = operator_arity!(arities self $({ $($arg: $argty),* })? $($ann)*);
+	    let (_params, _results) = arities.expect("instruction has unknown arities");
+
+	    define_visit_folded!(get_block_type $op blockty $($($arg)*)?);
+
+	    let use_color = self.printer.result.supports_async_color();
+	    let mut buf_color = PrintTermcolor(Ansi::new(Vec::new()));
+	    let mut buf_nocolor = PrintTermcolor(NoColor::new(Vec::new()));
+	    let internal_config = Config { print_offsets: false,
+					   print_skeleton: false,
+					   name_unnamed: self.printer.config.name_unnamed,
+					   fold_instructions: false };
+	    let mut internal_printer = Printer { config: &internal_config,
+						 result: if use_color { &mut buf_color } else { &mut buf_nocolor },
+						 nesting: self.printer.nesting,
+						 line: self.printer.line,
+						 group_lines: Vec::new(),
+						 code_section_hints: Vec::new() };
+
+	    let mut op_printer = PrintOperator::new(&mut internal_printer, self.state, self.operator_state);
```

Comment:
As you've probably grown accustomed to at this point I find that code in macros is best minimized where possible, and the above bits look like they might be ripe for extraction into a separate function. Perhaps something like:

```rust
fn with_op_printer<R>(&mut self, f: impl FnOnce(PrintOperator<'some, 'lifetimes, Here, Unsure>) -> R) -> R {
    // ..
}
```

which should help keep the amount of stuff in this macro pretty small hopefully

Reply:
Ack -- will take a stab at it.

Reply:
I took your commit from ebd6de2232b786aa364df0319ce33f4803774c99.